format/postgres/common/pg_heap/postgres/pg_heap.go

```diff
@@ -0,0 +1,408 @@
+	HEAP_HOT_UPDATED  = 0x4000 /* tuple was HOT-updated */
+	HEAP_ONLY_TUPLE   = 0x8000 /* this is heap-only tuple */
+)
+
+// type = struct PageHeaderData
+/*    0      |     8 */ // PageXLogRecPtr pd_lsn;
+/*    8      |     2 */ // uint16 pd_checksum;
+/*   10      |     2 */ // uint16 pd_flags;
+/*   12      |     2 */ // LocationIndex pd_lower;
+/*   14      |     2 */ // LocationIndex pd_upper;
+/*   16      |     2 */ // LocationIndex pd_special;
+/*   18      |     2 */ // uint16 PdPagesizeVersion;
+/*   20      |     4 */ // TransactionId pd_prune_xid;
+/*   24      |     0 */ // ItemIdData pd_linp[];
+//
+/* total size (bytes):   24 */
+
+// type = struct PageXLogRecPtr {
+/*    0      |     4 */ // uint32 xlogid;
+/*    4      |     4 */ // uint32 xrecoff;
+
+/* total size (bytes):    8 */
+
+// typedef uint16 LocationIndex;
+// #define SizeOfPageHeaderData (offsetof(PageHeaderData, pd_linp))
+
+// type = struct HeapTupleHeaderData {
+/*    0      |    12 */ // union {
+/*                12 */ //     HeapTupleFields t_heap;
+/*                12 */ //     DatumTupleFields t_datum;
+//
+/* total size (bytes):   12  */
+/* } t_choice;             //
+/*   12      |     6 */ // ItemPointerData t_ctid;
+/*   18      |     2 */ // uint16 t_infomask2;
+/*   20      |     2 */ // uint16 t_infomask;
+/*   22      |     1 */ // uint8 t_hoff;
+/*   23      |     0 */ // bits8 t_bits[];
+/* XXX  1-byte padding  */
+//
+/* total size (bytes):   24 */
+const SizeOfHeapTupleHeaderData = 24
+
+// type = struct HeapTupleFields {
+/*    0      |     4 */ // TransactionId t_xmin;
+/*    4      |     4 */ // TransactionId t_xmax;
+/*    8      |     4 */ // union {
+/*                 4 */ //    CommandId t_cid;
+/*                 4 */ //    TransactionId t_xvac;
+//                         } t_field3;
+/*                      total size (bytes):    4 */
+//
+/* total size (bytes):   12 */
+
+// type = struct DatumTupleFields {
+/*    0      |     4 */ // int32 datum_len_;
+/*    4      |     4 */ // int32 datum_typmod;
+/*    8      |     4 */ // Oid datum_typeid;
+//
+/* total size (bytes):   12 */
+
+// type = struct ItemPointerData {
+/*    0      |     4 */ // BlockIdData ip_blkid;
+/*    4      |     2 */ // OffsetNumber ip_posid;
+//
+/* total size (bytes):    6 */
+
+type Heap struct {
+	Args format.PostgresHeapIn
+
+	// current Page
+	Page *HeapPage
+	// Page special data
+	Special *PageSpecial
+
+	// current tuple
+	Tuple *TupleD
+
+	DecodePageHeaderData func(page *HeapPage, d *decode.D)
+	DecodePageSpecial    func(heap *Heap, d *decode.D)
+}
+
+type PageSpecial struct {
+	// pgproee
+	PdXidBase   uint64 // 8 TransactionId pd_xid_base;
+	PdMultiBase uint64 // 8 TransactionId pd_multi_base;
+	PdPruneXid  uint64 // 4 ShortTransactionId pd_prune_xid;
+	PdMagic     uint64 // 4 uint32 pd_magic;
+}
+
+type TupleD struct {
+	IsMulti bool
+}
+
+func Decode(heap *Heap, d *decode.D) any {
+	d.SeekAbs(0)
```

Comment:
Possible to make it up to the caller to make sure the position is correct? makes the function more independent.

The flavour thing made me think that maybe this could be implemented similar to how the network decoders interact in fq, they use internal format groups (tcp_stream, udp_datagram etc) that each decoder register to and then a decoder fill in a struct (port number etc) that it passed to the group and then it's up to each decoder to check port number etc which makes them decoupled. But the would require each postgres flavour to be a separate format.

Something to think about, i don't know myself what i would prefer. But it would make each decoder by design very isolated and indepdent which is nice.

Reply:
I removed SeekAbs(0). It's not used.